Entity Card: Card Addition

zsarnett opened the pull request:
## Proposed change

Entity Card - Similar to Sensor Card, except you do not have to use a Sensor Entity. This PR is similar to many of the current PRs. We are moving away from having to use extra sensors to get Attribute Data

The card will allow users to display state data or attribute data visually in a card

![image](https://user-images.githubusercontent.com/18730868/75192978-57fea700-5723-11ea-8d99-d239a52b2cd5.png)


## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml
- cards:
   - attribute: brightness
      default: '0'
      entity: light.bedroom
      type: entity
      unit: '%'
   - attribute: current_temperature
      entity: climate.ecobee
      type: entity
      unit: °F
   - entity: sensor.processor_use
      name: Processor
      type: entity
  type: horizontal-stack
- type: entity
   entity: weather.home
- type: entity
   entity: cover.kitchen_window
- cards:
    - type: entity
       entity: device_tracker.demo_paulus
    - type: entity
       entity: device_tracker.demo_anne_therese
  type: horizontal-stack
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->
[docs-repository]: https://github.com/home-assistant/home-assistant.io

zsarnett commented on translations/en.json:
```diff
@@ -1955,6 +1955,9 @@
                             "show_header_toggle": "Show Header Toggle?",
                             "toggle": "Toggle entities."
                         },
+                        "entity": {
```
I am not sure I did Translations correctly. I can't seem to get them to show in the frontend. Just blanks..

balloob replied:
Change `src/translations/en.json`

zsarnett replied:
smh changed the wrong translations...

zsarnett committed:
Review to changed Src translations

balloob commented on src/panels/lovelace/cards/hui-entity-card.ts:
```diff
@@ -0,0 +1,250 @@
+          <span id="value"
+            >${this._config.attribute
+              ? attributeValue
+              : computeStateValue(
```
The downside of this approach imo is that we put a lot of work on the users, instead of us doing the right thing. This method is a step in the right direction, but it's still not aggressive enough. We can do better.

Let's start with creating pre-made configurations.

You pick a climate entity: offer the user a dropdown if they want to see the HVAC Mode, action, current temperature, current humidity (if available) or target temp.

Then there is also a custom mode where it's all whatever. Let's start with just the custom mode in this PR: give us an attribute or we just render the state. No data processing there. If it's state we're rendering, we can default to unit to the unit_of_measurement attribute, otherwise empty.

zsarnett replied:
So you are saying to remove any processing to make the data look nice. Ex. `off` -> `Off`
And remove the ability to give a Unit of measurement

balloob replied:
Remove all the magic.

balloob commented on src/common/entity/compute_state_value.ts:
```diff
@@ -0,0 +1,81 @@
+      );
+      display = formatDateTime(date, language);
+    }
+  } else if (domain === "zwave") {
```
Did you copy paste this function from somewhere?

zsarnett replied:
Uh yes. 🙂. State Display function appended the UOM which I didn't want

balloob replied:
State and unit are connected. You can't just do one and have another function replicating the same logic just to see what the unit is.

zsarnett replied:
They are connected, sure. But the ability to change that UOM is not foreign

zsarnett committed:
Reviews

zsarnett committed:
side by side theme

bramkragten commented:
This is almost if not completely the same as a sensor card without a graph? Why not combine the 2?

bramkragten commented:
We have `prefix` and `suffix` in other cards should we align that? Or...?

balloob commented:
Please put this on hold until we have shipped media player. One beast at a time.

zsarnett commented:
@balloob Where are we going with this :) I am still unsure. Maybe some time we talk this through? It sounds like a cool idea you have but not sure the full scope.

balloob commented:
Let's jump on a call tomorrow. Maybe include Bram too.

zsarnett commented:
Sounds good. Let me know a time when you are available. I have no other meetings today

bramkragten commented:
Can you guys make it in 45 min?

zsarnett commented:
@bramkragten I can make it. 12 PM EST. @balloob 9:00 AM for you?

zsarnett committed:
Allow user to specify unit

zsarnett committed:
Add unit back update for headerfooter and editor

zsarnett committed:
Clean

zsarnett commented:
So from our conversation this morning:

- I added the ability to configure a unit. To make this a fully manual card as discussed
- Added the ability to set a header and footer
- No magic is involved

Future PRs will include

- Making graph an option for the footer/header
- Converting the sensor card to an entity card with a footer

zsarnett commented:
Should Default be an option here? We talked about this a bit in discord but what should we do here if, for example, the attribute doesn't exist if the state is off

Should we fall back to the State?

Should we render something that says no data?

With this current version of this card is all manual. Maybe we do have a default option. In the future we can set defaults automatically. Example: Hvac humidity - If the hvac is off and it doesn't exists we render Off.

balloob commented:
If it's all manual, we should not do magic.
 - If state is unknown or unavailable, we can just render that.
 - If attribute doesn't exist, treat it as unknown. Do not render unit.
 - Otherwise render as usual.

zsarnett commented:
So if the attribute doesn't exist. Render the value as "Unknown" (Localized)?

zsarnett committed:
Unavailable for attribute that doesn't exist

zsarnett commented:
I also still feel like we should localize the state. I wouldn't consider Localizing into a  friendly state name magic. Aka (not_home -> Not Home) also this removes all other languages. So now they see the English version of the state since that's how we developed it

zsarnett committed:
fix merge

zsarnett committed:
Fix from rebasing

zsarnett commented:
@balloob I left a few comments above about the recent changes. Would you be able to review sometime?

zsarnett commented:
Also, What should the documentation look like? Not sure how this card should be marketed anymore

balloob commented on src/panels/lovelace/cards/hui-entity-card.ts:
```diff
@@ -0,0 +1,270 @@
+      (!this._config.attribute &&
+        stateObj.state !== UNKNOWN &&
+        stateObj.state !== UNAVAILABLE) ||
+      (this._config.attribute && stateObj.attributes[this._config.attribute]);
```
Little easier to read if you use a ternary statement.

balloob commented on src/panels/lovelace/cards/hui-entity-card.ts:
```diff
@@ -0,0 +1,270 @@
+          >${showUnit
+            ? html`
+                <span id="measurement"
+                  >${this._config.unit ||
```
This check means that now if you want to show no unit, you need to put in a space.

zsarnett replied:
What would you suggest here?

```ts
${this._config.unit || this._config.attribute ? "" : stateObj.attributes.unit_of_measurement}
```

zsarnett replied:
A config variable for show unit?

balloob replied:
`"unit" in this._config`

zsarnett replied:
Doesnt the above do the same? This would just be an extra step?

```ts
"unit" in this._config ? this._config.unit : this._config.attribute ? "" : stateObj.attributes.unit_of_measurement
```

because if the attribute is set then we don't want the state measurement

zsarnett replied:
I guess maybe 

```ts
!("unit" in this._config) && this._config.attribute ? "" : this._config.unit || stateObj.attributes.unit_of_measurement
```

balloob commented on src/panels/lovelace/cards/hui-entity-card.ts:
```diff
@@ -0,0 +1,270 @@
+    const element = createHeaderFooterElement(conf);
+    if (this.hass) {
+      element.hass = this.hass;
+    }
```
Why this check?

zsarnett replied:
This is straight from the entities card. Should we not check this here?

balloob commented on src/panels/lovelace/cards/hui-entity-card.ts:
```diff
@@ -0,0 +1,270 @@
+    conf: LovelaceHeaderFooterConfig,
+    className: string
+  ): TemplateResult {
+    const element = createHeaderFooterElement(conf);
```
Now we're throwing away and re-creating this element at every render. And since `hass` is a property, that's a lot. You should create this element in `setConfig` and update it in `updated` with new `hass`

zsarnett replied:
We should probably update the entities card then, no?

balloob replied:
Nope, the entities card has a setter for `hass`. It only re-renders when new config arrives, otherwise it manually looks at the DOM elements and finds the ones that need `hass` updated.

balloob replied:
You cannot do that here because you are rendering a template based on hass-data. Entities card just needs to pass it along to the rows/header/footer

zsarnett replied:
Got it. Makes sense

balloob commented on src/panels/lovelace/cards/hui-entity-card.ts:
```diff
@@ -0,0 +1,270 @@
+
+      ha-card {
+        display: flex;
+        flex-direction: column;
```
Why are both column ?

balloob commented on src/panels/lovelace/cards/hui-entity-card.ts:
```diff
@@ -0,0 +1,270 @@
+        font-weight: 500;
+        max-height: 1.4rem;
+        top: 2px;
+        opacity: 0.8;
```
Isn't this going to mess with all the colors that people pick ?

balloob replied:
And since it's inside `.name`, which also has an opacity, does it mean that `.name > span` is shown with `0.64` opacity?

zsarnett replied:
Probably but this is how the sensor card has been since creation. I like the greyed look of the header. Gives it depth and separation from the value

zsarnett replied:
I removed the span in name as it was unneeded

balloob replied:
That's why we have secondary text color as a variable.

balloob commented on src/panels/lovelace/cards/hui-entity-card.ts:
```diff
@@ -0,0 +1,270 @@
+      }
+
+      .icon {
+        color: var(--paper-item-icon-color, #44739e);
```
paper-item ?

zsarnett replied:
Not sure what the Variable should be here?

Seems this is what is used

balloob replied:
@bramkragten what should we use?

bramkragten replied:
`--state-icon-color`

balloob commented on src/panels/lovelace/cards/hui-entity-card.ts:
```diff
@@ -0,0 +1,270 @@
+        margin: 0 16px 16px;
+      }
+
+      #value {
```
Let's not use IDs, just classes.

balloob commented on src/panels/lovelace/cards/hui-entity-card.ts:
```diff
@@ -0,0 +1,270 @@
+        overflow: hidden;
+        text-overflow: ellipsis;
+        -webkit-line-clamp: 1;
+        -webkit-box-orient: vertical;
```
You only need to use line-clamp if you want multi-line overflow ellipsis effect. Since this is 1 line, you can remove this.

balloob commented on src/panels/lovelace/cards/hui-entity-card.ts:
```diff
@@ -0,0 +1,270 @@
+        background: var(--divider-color);
+      }
+
+      .flex {
```
Not a fan of extracting this. It makes reading the CSS a bit more complicated as it now has a dependency on the right classes being set in the HTML.

balloob commented on src/panels/lovelace/cards/hui-entity-card.ts:
```diff
@@ -0,0 +1,270 @@
+        display: flex;
+        min-width: 0;
+        opacity: 0.8;
+        position: relative;
```
Why is it position relative ?

balloob commented on src/panels/lovelace/cards/hui-entity-card.ts:
```diff
@@ -0,0 +1,270 @@
+      .name {
+        align-items: center;
+        display: flex;
+        min-width: 0;
```
Why this?

zsarnett replied:
Seems like I need to go through a rewrite the CSS for this card. This was copied from the sensor card so most of these things you are reviewing I don't know either :)

balloob commented on src/translations/en.json:
```diff
@@ -2145,7 +2149,8 @@
           "entity_not_found": "Entity not available: {entity}",
-          "entity_non_numeric": "Entity is non-numeric: {entity}"
+          "entity_non_numeric": "Entity is non-numeric: {entity}",
+          "attribute_not_found": "Attribute not available: {attribute}"
```
This one already exists 4 lines above it?

balloob replied:
I also don't see it being used?

balloob commented on src/panels/lovelace/cards/hui-entity-card.ts:
```diff
@@ -0,0 +1,270 @@
+        align-self: flex-end;
+        display: inline-block;
+        font-size: 1.3rem;
+        line-height: 1.2em;
```
We're mixing rem and em.

balloob commented on src/panels/lovelace/cards/hui-entity-card.ts:
```diff
@@ -0,0 +1,270 @@
+        display: inline-block;
+        font-size: 1.3rem;
+        line-height: 1.2em;
+        margin-top: 0.1em;
```
Why do this in em? Are you trying to make this card grow proportionally when the user increases font size?

zsarnett replied:
I didn't make this decision 🙂

zsarnett committed:
reviews

zsarnett committed:
Localize State

zsarnett committed:
fix for localize

zsarnett committed:
css

balloob commented on src/panels/lovelace/cards/hui-entity-card.ts:
```diff
@@ -0,0 +1,235 @@
+
+    if (this._config.footer) {
+      this._footerElement = createHeaderFooterElement(this._config.footer);
+    }
```
Make sure you unset it, or else removing a footer in a config update does nothing.

balloob commented on src/panels/lovelace/cards/hui-entity-card.ts:
```diff
@@ -0,0 +1,235 @@
+  }
+
+  public getCardSize(): number {
+    return 2;
```
```suggestion
    return 2 + (this._config.footer ? 1 : 0);
```

balloob replied:
Or maybe ask the footer for size ? (although I don't think we implemented that)

zsarnett replied:
This is probably something we should implement in the future as the graph footer is much bigger than the buttons. But for now.

balloob commented on src/panels/lovelace/cards/hui-entity-card.ts:
```diff
@@ -0,0 +1,235 @@
+    }
+
+    const showUnit = this._config.attribute
+      ? Boolean(stateObj.attributes[this._config.attribute])
```
This will filter out if the value is falsey, like 0. 
```suggestion
      ? this._config.attribute in stateObj.attributes
```

balloob commented on src/panels/lovelace/cards/hui-entity-card.ts:
```diff
@@ -0,0 +1,235 @@
+        <div class="info">
+          <span class="value"
+            >${this._config.attribute
+              ? stateObj.attributes[this._config.attribute] ||
```
handle attribute being `0`

zsarnett committed:
reviews

zsarnett committed:
Break the rules

balloob commented:
Ok to merge when docs added !
